File: frontend/src/api/shopping.ts
import { apiRequest } from "./client";

export type ShoppingQuantityComponent = {
  quantity: string;
  unit_symbol: string;
};

export type ShoppingSourceContribution = {
  meal_plan_item_id: number;
  date: string;
  meal_slot: "lunch" | "dinner";
  dish_name: string;
  recipe_variant_name: string | null;
  quantity: string;
  unit_symbol: string;
  optional: boolean;
};

export type ShoppingPlannedMeal = {
  meal_plan_item_id: number;
  date: string;
  meal_slot: "lunch" | "dinner";
  dish_name: string;
  recipe_variant_name: string | null;
};

export type ShoppingListItem = {
  id: number | null;
  ingredient_id: number;
  display_name: string;
  quantity: string;
  unit_id: number;
  unit_symbol: string;
  category: string;
  checked: boolean;
  approximate: boolean;
  optional: boolean;
  source_meal_plan_item_ids: number[];
  source_contributions: ShoppingSourceContribution[];
  raw_components: ShoppingQuantityComponent[];
};

export type ShoppingList = {
  id: number | null;
  from_date: string;
  to_date: string;
  status: "draft" | "active" | "completed" | "archived";
  exclude_pantry: boolean;
  items: ShoppingListItem[];
  planned_meals: ShoppingPlannedMeal[];
};

export async function previewShoppingList(
  token: string,
  params: { from: string; days: number; exclude_pantry: boolean },
): Promise<ShoppingList> {
  const search = new URLSearchParams({
    from: params.from,
    days: String(params.days),
    exclude_pantry: String(params.exclude_pantry),
  });
  return apiRequest<ShoppingList>(`/api/shopping-list?${search.toString()}`, { token });
}

export async function createShoppingList(
  token: string,
  body: { from_date: string; days: number; exclude_pantry: boolean },
): Promise<ShoppingList> {
  return apiRequest<ShoppingList>("/api/shopping-lists", {
    method: "POST",
    token,
    body: JSON.stringify(body),
  });
}

export async function fetchShoppingList(token: string, shoppingListId: number): Promise<ShoppingList> {
  return apiRequest<ShoppingList>(`/api/shopping-lists/${shoppingListId}`, { token });
}

export async function updateShoppingListItem(
  token: string,
  itemId: number,
  body: { checked: boolean },
): Promise<ShoppingListItem> {
  return apiRequest<ShoppingListItem>(`/api/shopping-list-items/${itemId}`, {
    method: "PUT",
    token,
    body: JSON.stringify(body),
  });
}

File: backend/mealroulette/schemas/shopping.py
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field

from mealroulette.models.enums import MealSlot, ShoppingListStatus


class ShoppingQuantityComponent(BaseModel):
    quantity: Decimal
    unit_symbol: str


class ShoppingSourceContribution(BaseModel):
    meal_plan_item_id: int
    date: date
    meal_slot: MealSlot
    dish_name: str
    recipe_variant_name: str | None = None
    quantity: Decimal
    unit_symbol: str
    optional: bool = False


class ShoppingSourceMeal(BaseModel):
    meal_plan_item_id: int
    date: date
    meal_slot: MealSlot
    dish_name: str
    recipe_variant_name: str | None = None


class ShoppingPlannedMeal(BaseModel):
    meal_plan_item_id: int
    date: date
    meal_slot: MealSlot
    dish_name: str
    recipe_variant_name: str | None = None


class ShoppingListItemPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int | None = None
    ingredient_id: int
    display_name: str
    quantity: Decimal
    unit_id: int
    unit_symbol: str
    category: str
    checked: bool = False
    approximate: bool = False
    optional: bool = False
    source_meal_plan_item_ids: list[int] = Field(default_factory=list)
    source_meals: list[ShoppingSourceMeal] = Field(default_factory=list)
    source_contributions: list[ShoppingSourceContribution] = Field(default_factory=list)
    raw_components: list[ShoppingQuantityComponent] = Field(default_factory=list)
    created_at: datetime | None = None
    updated_at: datetime | None = None


class ShoppingListPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int | None = None
    from_date: date
    to_date: date
    status: ShoppingListStatus = ShoppingListStatus.active
    exclude_pantry: bool = True
    items: list[ShoppingListItemPublic] = Field(default_factory=list)
    planned_meals: list[ShoppingPlannedMeal] = Field(default_factory=list)
    created_at: datetime | None = None
    updated_at: datetime | None = None


class ShoppingListCreateRequest(BaseModel):
    from_date: date
    days: int = Field(ge=1, le=14)
    exclude_pantry: bool = True


class ShoppingListItemUpdateRequest(BaseModel):
    checked: bool | None = None
